### DISCUSSION TITLE
Extending Obsidian query syntax

### DISCUSSION THREAD

**Discussion by srid** [srid/emanote#48] [Ideas]
Main issue with sub-tasks: https://github.com/srid/emanote/issues/10

https://help.obsidian.md/Plugins/Search#Embed+search+results

- [x] Add "timeline" class for listing with date
- [x] Support `path:./*` for listing notes under the current "directory zettel" (i.e. if used in `foo.md` list notes immediately under `foo/`).

Related: https://github.com/srid/emanote/discussions/106

**Comment by srid** [comment]
@xygn I noticed that you use Obsidian. How is the `path` query working for you? Both in Obsidian & Emanote. Any thoughts on this feature as being implemented in Emanote? 

I believe Obsidian does a simple sub-string search, but Emanote supports pattern-matching (using `*`) which, IIUC, Obsidian does not support.

**Reply by xygn** [reply]
It's working great! In Obsidian, as you said, it brings out the global sub-string matches. Thus if you just specify a substring it lists notes that contains the substring titles globally too but Emanote doesn't do the same. 

I [tested ](https://raw.githubusercontent.com/xygn/Notes/master/class/Core/Pathology.md) out various path query types, it looks like Emanote requires a wildcard to match path, and just specifying the directory

In contrast, Emanote does pattern-matching, but Obsidian does not support wildcard pattern-matching, as you said. I'm not sure if cross-compatibility could be done, maybe you could support Obsidian-style querying in addition to the current pattern-matching that Emanote supports.

Here's the difference in parallel:
![image](https://user-images.githubusercontent.com/63365609/124240188-23c9a080-db38-11eb-8565-23ec06f07a5f.png)


I think it's a great feature, and an improvement that could be helpful: if there's no pre-existing markdown file for a folder `foo`, could we have it autogenerate a webpage by default for directories without the need for a `foo.md`, instead of showing a blank page when there's no `foo.md`? 

And this can be overridden when there is a `foo.md` under the parent directory of `foo`.

Also is there anyway sub-string search could possibly be implemented? Such as:
>       ```query
>         Acute
>      ```

Obsidian:
![image](https://user-images.githubusercontent.com/63365609/124246635-b9682e80-db3e-11eb-9aae-24b5cd414d2a.png)
Emanote:
![image](https://user-images.githubusercontent.com/63365609/124247890-f2ed6980-db3f-11eb-879b-86c48fe8f644.png)

Maybe a search box would be better instead of this. And afaik, Obsidian Publish doesn't support it either.

**Reply by srid** [reply]
> In contrast, Emanote does pattern-matching, but Obsidian does not support wildcard pattern-matching, as you said. I'm not sure if cross-compatibility could be done, maybe you could support Obsidian-style querying in addition to the current pattern-matching that Emanote supports.

I've slightly changed the `path:` query now. If you use `path:/Foo/Bar` it will list files under `/Foo/Bar` (note: it is top-level "Foo" directory). But if you use `path:Foo/Bar` (note the absence of the leading slash), it will be equivalent to `path:**/Foo/Bar/**`. But I suggest not using the latter syntax yet ... until full-text search (see below) is implemented, because then we can make the non-`/`-at-the-front value such as `path:Foo/Bar` to look for sub-string in the path.

Note that to mimic your screenshot in Emanote, you can try: `path:**/General/**` (or just `path:General` for now). You can use any sub-path, eg: `path:**/Pathology/General/**` (or `path:Pathology/General`).

I will try to be compatible with Obsidian to the extent it makes sense, but there is no guarantee of 100% compatibility.

> if there's no pre-existing markdown file for a folder `foo`, could we have it autogenerate a webpage by default for directories without the need for a `foo.md`, instead of showing a blank page when there's no `foo.md`?

It already does that. Does it not for you? In today's commit, I made it display the folder listing as well.

> Also is there anyway sub-string search could possibly be implemented?

For full-text search, see https://github.com/srid/emanote/issues/102

**Reply by xygn** [reply]
> Note that to mimic your screenshot in Emanote, you can try: `path:**/General/**` (or just `path:General` for now). You can use any sub-path, eg: `path:**/Pathology/General/**` (or `path:Pathology/General`).
>
> I will try to be compatible with Obsidian to the extent it makes sense, but there is no guarantee of 100% compatibility.

 The present implementation has already made it mostly compatible. Thanks! 

> It already does that. Does it not for you? In today's commit, I made it display the folder listing as well.

Yes, I meant the folder listing by default. It's perfect now. This makes mobile navigation a bit easier too.

**Comment by srid** [comment]
See https://github.com/srid/emanote/discussions/106 for a general version of this syntax.
